### PR TITLE
fix(tianmu):query with UNION ALL return result set error.(#854)

### LINKED ISSUE #854: bug: query with UNION ALL return result set error
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

The wrong result is as follows:

```shell
mysql> select a from t1;
+----------+
| a        |
+----------+
| -8388606 |
|       -1 |
|        0 |
|        1 |
|  8388607 |
+----------+
5 rows in set (0.00 sec)

mysql> select a from t2;
+----------------------+
| a                    |
+----------------------+
| -9223372036854775806 |
|                   -1 |
|                    0 |
|                    1 |
|  9223372036854775807 |
+----------------------+
5 rows in set (0.00 sec)

mysql> SELECT a FROM t1 UNION ALL SELECT a FROM t2;
+----------+
| a        |
+----------+
| -8388606 |
|       -1 |
|        0 |
|        1 |
|  8388607 |
|        2 |
|       -1 |
|        0 |
|        1 |
|       -1 |
+----------+
10 rows in set (0.00 sec)
```

### Expected behavior

```shell
mysql> SELECT a FROM t1 UNION ALL SELECT a FROM t2;
+----------------------+
| a                    |
+----------------------+
|             -8388606 |
|                   -1 |
|                    0 |
|                    1 |
|              8388607 |
| -9223372036854775806 |
|                   -1 |
|                    0 |
|                    1 |
|  9223372036854775807 |
+----------------------+
10 rows in set (0.00 sec)
```

### How To Reproduce

```sql
CREATE TABLE t1 (a MEDIUMINT);
INSERT INTO t1 VALUES (CAST(0xFFFFFFFFFF800000 AS SIGNED)+2),(-1),(0),(1),(0x7FFFFF);
CREATE TABLE t2 (a BIGINT);
INSERT INTO t2 VALUES (CAST(0x8000000000000000 AS SIGNED)+2),(-1),(0),(1),(0x7FFFFFFFFFFFFFFF);
SELECT a FROM t1 UNION ALL SELECT a FROM t2;
```

### Environment

```shell
root@ub01:~# /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: 86a7685dc
        Last commit time: Date:   Sat Oct 22 03:38:32 2022 +0800
        Build time: Date: Sat 22 Oct 2022 02:14:05 PM UTC
root@ub01:~# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by DandreChen**
## Summary about this PR

Issue Number: close #854 

[summary]
Modify the query logic executed by the union to make the result: go through the prepare() process

## Tests Check List

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/engine_execute.cpp [hunk 1]
```diff
@@ -456,12 +456,14 @@ QueryRouteTo Engine::Execute(THD *thd, LEX *lex, Query_result *result_output, SE
...
       rct = current_txn_->GetTableByPathIfExists(table_path);
     }
-    if (unit_for_union != nullptr && !unit_for_union->is_prepared()) {
-      int res = result_output->prepare(unit_for_union->item_list, unit_for_union);
-      if (res) {
-        TIANMU_LOG(LogCtl_Level::ERROR, "Error: Unsupported UNION");
-        my_message(ER_UNKNOWN_ERROR, "Tianmu: unsupported UNION", MYF(0));
-        throw ReturnMeToMySQLWithError();
+    if (unit_for_union != nullptr) {
+      if (lex->sql_command != SQLCOM_CREATE_TABLE) {  //  for exclude CTAS
```
Comment (RingsC) [inline storage/tianmu/core/engine_execute.cpp]: 1: Why do we exclude CTAS from this.
2: can be rewrite as `if (unit_for_union != nullptr && lex->sql_command != SQLCOM_CREATE_TABLE )`

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/engine_execute.cpp (hunk 1 above)
Comment (DandreChen) [reply storage/tianmu/core/engine_execute.cpp]: if not exclude CATS from this, db will crash, as below sql:
```
 CREATE TABLE t1 (f1 VARCHAR(255) CHARACTER SET utf8) ENGINE=TIANMU;
 CREATE TABLE t2 AS SELECT LEFT(f1,171) AS f2 FROM t1 UNION SELECT LEFT(f1,171) AS f2 FROM t1;
```

Comment (DandreChen) [reply storage/tianmu/core/engine_execute.cpp:460]: The purpose of this writing is to create a correct table structure from union select

Comment (DandreChen) [reply storage/tianmu/core/engine_execute.cpp]: have rewrited

**Review by RingsC — APPROVED** [review]
LGTM

**Review by hustjieke — APPROVED** [review]
Nice job! LGTM

**Pull request merged**